lib/code-caller-docker.js

```diff
@@ -0,0 +1,743 @@
+    // macOS
+    await spawnAsync('bindfs', [directory, mountpoint]);
+  } else if (process.platform === 'linux') {
+    // linux
+    await spawnAsync('mount', ['--bind', directory, mountpoint]);
+  } else {
+    throw new Error(`Unsupported platform for bind mounts: ${process.platform}`);
+  }
+}
+
+/**
+ *
+ * @param {string} mountpoint
+ */
+async function removeBindMount(mountpoint) {
+  if (process.platform === 'darwin' || process.platform === 'linux') {
+    await spawnAsync('umount', [mountpoint]);
+  } else {
+    throw new Error(`Unsupported platform for bind mounts: ${process.platform}`);
+  }
+}
+
+class DockerCaller {
+  constructor(options = { questionTimeoutMilliseconds: 5000 }) {
+    /** @type {CallerState} */
+    this.state = CREATED;
+    this.uuid = uuidv4();
+
+    this.debug('enter constructor()');
+
+    this.container = null;
+    this.callback = null;
+    this.timeoutID = null;
+    this.callCount = 0;
+    this.ensureChildMutex = new Mutex();
+    this.ensureDirectoryMutex = new Mutex();
+
+    this.options = options;
+
+    // this will accumulate stdout from the container
+    this.output = '';
+
+    // for error logging
+    this.lastCallData = null;
+
+    this.coursePath = null;
+
+    this._checkState();
+
+    this.debug(`exit constructor(), state: ${String(this.state)}, uuid: ${this.uuid}`);
+  }
+
+  /**
+   * Wrapper around `debug` that automatically includes UUID and the caller state.
+   *
+   * @param {string} message
+   */
+  debug(message) {
+    const paddedState = this.state.toString().padEnd(15);
+    debug(`[${this.uuid} ${paddedState}] ${message}`);
+  }
+
+  /**
+   * Wrapper around `createBindMount` that includes instance-specific logs.
+   *
+   * @param {string} directory
+   * @param {string} mountpoint
+   */
+  async createBindMount(directory, mountpoint) {
+    this.debug(`creating bind mount for ${directory} at ${mountpoint}`);
+    await createBindMount(directory, mountpoint);
+    this.debug(`created bind mount for ${directory} at ${mountpoint}`);
+  }
+
+  /**
+   * Wrapper around `removeBindMount` that includes instance-specific logs.
+   *
+   * @param {string} mountpoint
+   */
+  async removeBindMount(mountpoint) {
+    this.debug(`removing bind mount at ${mountpoint}`);
+    await removeBindMount(mountpoint);
+    this.debug(`removed bind mount at ${mountpoint}`);
+  }
+
+  /**
+   * Allows this caller to prepare for execution of code from a particular
+   * course.
+   *
+   * @param {string} coursePath
+   * @param {(err: Error | null | undefined) => void} callback
+   */
+  prepareForCourse(coursePath, callback) {
+    util.callbackify(async () => {
+      if (this.coursePath && this.coursePath === coursePath) {
+        // Same course as before; we can reuse the existing setup
+        return;
+      }
+
+      this.coursePath = coursePath;
+
+      if (config.workersExecutionMode === 'container') {
+        await this._ensureDirectoryIfNeeded();
```

Comment:
`IfNeeded` is outdated nomenclature since a directory is _always_ needed - should rename to just `ensureDirectory`.

Reply:
Done in fe93c5c.